### ISSUE TITLE
[DataGrid] Performance issues with large dataset (1m+)

### ISSUE THREAD

**Issue by TikoAbelyan** [mui/mui-x#10896]
### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: large dataset
**Order ID**: 77939

**Comment by alexfauquette** [comment]
Could it be linked to the size of the scroll container?

> *unlimited: Browsers set a limit on the number of pixels a scroll container can host: 17.5 million pixels on Firefox, 33.5 million pixels on Chrome, Edge, and Safari. A [reproduction](https://codesandbox.io/s/beautiful-silence-1yifo?file=/src/App.js).

https://mui.com/x/react-data-grid/virtualization/

What's your use case for having to display that many rows at once?

**Comment by romgrk** [comment]
Please provide a minimal reproduction test case with the latest version. This would help a lot 👷.
A live example would be perfect. [This codesandbox.io template](https://codesandbox.io/s/github/mui/mui-x/tree/master/templates/x-data-grid?file=/src/demo.tsx) may be a good starting point. Thank you!

**Comment by TikoAbelyan** [comment]
Hey @romgrk , thanks for your help.

```
  const loadServerRows = async (firstRowToRender: number) => {
    const currentPage = Math.floor(firstRowToRender / pageSize);
    const newData = await fetchPageData(currentPage);

    if (mounted.current) {
      apiRef.current.unstable_replaceRows(currentPage * pageSize, newData);
    }
  };
  
<DataGridPro
        columns={columns}
        rows={initialData}
        rowCount={totalCount} 
        apiRef={apiRef}
        loading={loading}
        className={classes.dataGrid}
        disableRowSelectionOnClick
        rowsLoadingMode="server"
        onFetchRows={(params) => loadServerRows(params.firstRowToRender)}
        experimentalFeatures={{
          lazyLoading: true,
        }}
        slots={{
          toolbar: () => (
            <div className={classes.headerTable}>
              <GridToolbar />
              <SliceSelector
                visualizationMode={visualizationMode}
                onSubmit={(i) => {
                  loadServerRows(i);
                  loadServerRows(i + pageSize + 1);
                  apiRef.current.scroll({
                    top: i * 52,
                  });
                }}
                tableRun
              />
            </div>
          ),
          footer: () => (
            <div>
                <Typography>Total rows: {totalCount}</Typography>
            </div>
          ),
        }}
      />
``` 

totalCount for now can be 1m or 2m+ etc.

I want to figure out how can I cut by 10000, I mean after scroll to 10.000 row it will reload and receive new data starting for 10001.

I understand that if my dataset length is 1m+ so height equal 52M and it began freeze, I want to fix by cutting the data in to smaller segments, do you have example or do you have any suggestion ? thx

**Comment by romgrk** [comment]
Hi @TikoAbelyan, can you please put the code in a *runnable codesandbox*, and share the link with us. It's important that we have a reproducible code example before we can help you. You can use the codesandbox template in my message above.

**Comment by TikoAbelyan** [comment]
Here is example sometimes it freeze, btw height now is `78.000.000px`

https://codesandbox.io/s/zealous-allen-d33m87?file=/src/demo.tsx

<img width="1500" alt="Screenshot 2023-11-07 at 18 53 13" src="https://github.com/mui/mui-x/assets/28678783/542f706a-553e-4fb1-9d60-108c48b38df0">

**Comment by romgrk** [comment]
It seems like the problem is there is not enough memory for the data. This doesn't seem like a datagrid specific problem.

**Comment by TikoAbelyan** [comment]
I agree , Is there variant for fix?, cutting for example rows by 100.000 or something else?

**Comment by romgrk** [comment]
Server side data, lazy loading. You can read more about those topics in the docs.
